deep_qa/common/params.py

```diff
@@ -21,74 +23,119 @@ def param(self, message, *args, **kws):
-def get_choice(params: Dict[str, Any], key: str, choices: List[Any], name: str=None):
+class Params(MutableMapping):
     """
-    Gets the value of `key` in the `params` dictionary, ensuring that the value is one of the given
-    choices.  `name` is an optional description for where a configuration error happened, if there
-    is one.
-
-    Note that this _pops_ the key from params, modifying the dictionary, consistent with how
-    parameters are processed in this codebase.
+    A Class representing a parameter dictionary with a history. Using this allows
+    exact reproduction of a parameter file used in an experiment from logs, even when
+    default values are used.
     """
 
-    value = pop_with_logging(params, key)
-    if value not in choices:
-        raise ConfigurationError(_get_choice_error_message(value, choices, name))
-    return value
-
-
-def get_choice_with_default(params: Dict[str, Any],
-                            key: str,
-                            choices: List[Any],
-                            default: Any=None,
-                            name: str=None):
-    """
-    Like get_choice, but with a default value.  If `default` is None, we use the first item in
-    `choices` as the default.
-    """
-    try:
-        return get_choice(params, key, choices, name)
-    except KeyError:
-        if default is None:
-            return choices[0]
-        if default not in choices:
-            raise ConfigurationError(_get_choice_error_message(default, choices, name))
-        return default
-
-
-def assert_params_empty(params: Dict[str, Any], class_name: str):
+    # This allows us to check for the presence of "None" as a default argument,
+    # which we require because we make a distinction bewteen passing a value of "None"
+    # and passing no value to the default parameter of "pop".
+    DEFAULT = object()
+
+    def __init__(self, params: Dict[str, Any], history: str=""):
+
+        self.params = params
+        self.history = history
+
+    @overrides
+    def pop(self, key: str, default: Any=DEFAULT) -> Union["Params", Any]:
+        """
+        Performs the functionality associated with dict.pop(key) but with parameter
+        logging. This is required because pop_with_default may receive a default value
+        of None, which means we can't check for it not being passed.
+        """
+        if default is self.DEFAULT:
+            value = self.params.pop(key)
+        else:
+            value = self.params.pop(key, default)
+        logger.param(key + " : " + str(value))
+        return self.__check_is_dict(key, value)
+
+    @overrides
+    def get(self, key: str, default: Any=DEFAULT) -> Union["Params", Any]:
+        """
+        Performs the functionality associated with dict.pop(key) but with parameter
+        logging. This is required because pop_with_default may receive a default value
+        of None, which means we can't check for it not being passed.
+        """
+        print(default)
```

Comment:
Ah sorry this should implement the `get` method of a dict, I just copied it over from `pop` and forgot to change it, thanks for spotting